File: Components/Strata.Client.Core/main/typescript/Presenter/SimpleAction.ts
import {IAction} from "./IAction";
import {CompletableObservable,ICompletionStage} from "strata.foundation.core/Concurrent";

export
class SimpleAction<M>
    implements IAction<M>
{
    private readonly key: string;
    private readonly action: (model: M) => M;

    constructor(key: string,action: (model: M) => M)
    {
        this.key = key;
        this.action = action;
    }

    getKey(): string
    {
        return this.key;
    }

    apply(model: M): ICompletionStage<M>
    {
        return CompletableObservable.fromResult(this.action(model));
    }

}

File: Components/Strata.Client.Core/src/main/java/strata/client/core/presenter/SimpleAction.java
//////////////////////////////////////////////////////////////////////////////
// LambdaAction.java
//////////////////////////////////////////////////////////////////////////////

package strata.client.core.presenter;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;
import java.util.function.Function;

public
class SimpleAction<M>
    implements IAction<M>
{
    private final Class<M>      itsKey;
    private final Function<M,M> itsAction;

    public
    SimpleAction(Class<M> key,Function<M,M> action)
    {
        itsKey = key;
        itsAction = action;
    }

    @Override
    public Class<M>
    getKey()
    {
        return itsKey;
    }

    @Override
    public CompletionStage<M>
    apply(M model)
    {
        return
            CompletableFuture.completedFuture(itsAction.apply(model));
    }
}

//////////////////////////////////////////////////////////////////////////////
